File: internal/recovery/recovery.go
package recovery

import (
	"bigdb/internal/memtable"
	"bigdb/internal/record"
	"bigdb/internal/wal"
)

func ReplayToMemtable(w *wal.WAL, m *memtable.Memtable) error {
	return w.ReplayAll(func(rec record.Record) error {
		if rec.IsDeleted {
			m.Delete(rec)
			return nil
		}
		m.Put(rec)
		return nil
	})
}

// func DebugReplaySummary(w *wal.WAL) error {
// 	count := 0
// 	return w.Replay(func(rec record.Record) error {
// 		count++
// 		_ = rec
// 		return nil
// 	})
// }

File: internal/recovery/recovery.cpp
#include "recovery.h"

namespace bigdb::recovery {

void ReplayToMemtable(const wal::WAL& wal, memtable::Memtable& memtable) {
    auto records = wal.Replay();
    for (const auto& rec : records) {
        if (rec.isDeleted) {
            memtable.Delete(rec);
        } else {
            memtable.Put(rec);
        }
    }
}

}  // namespace bigdb::recovery
